src/openssl_wrap.cpp

```diff
@@ -864,6 +864,58 @@ std::string _X509_NAME_ENTRY_get_data(X509_NAME_ENTRY *entry)
     return bio.flushToString();
 }
 
+#pragma mark -- Symmetric Encryption Interface --
+
+SSL_EVP_CIPHER_CTX_Ptr _EVP_CIPHER_CTX_new() {
+    return SSL_EVP_CIPHER_CTX_Ptr{OpensslCallPtr::callChecked(lib::OpenSSLLib::SSL_EVP_CIPHER_CTX_new)};
+}
+
+void _EVP_CIPHER_CTX_reset(EVP_CIPHER_CTX *ctx) {
+    OpensslCallIsOne::callChecked(lib::OpenSSLLib::SSL_EVP_CIPHER_CTX_reset, ctx);
+}
+
+
+void _EVP_CIPHER_CTX_ctrl(EVP_CIPHER_CTX *ctx, int type, int arg, void *ptr) {
+    OpensslCallIsOne::callChecked(lib::OpenSSLLib::SSL_EVP_CIPHER_CTX_ctrl, ctx, type, arg, ptr);
+}
+
+void _EVP_CipherInit_ex(EVP_CIPHER_CTX *ctx,
+                        const EVP_CIPHER* cipher,
+                        ENGINE* impl,
+                        const unsigned char* key,
+                        const unsigned char* iv,
+                        int enc) {
+    OpensslCallIsOne::callChecked(lib::OpenSSLLib::SSL_EVP_CipherInit_ex, ctx, cipher, impl, key, iv, enc);
+}
+
+void _EVP_CipherUpdate(EVP_CIPHER_CTX * ctx,
+                       unsigned char * outm,
+                       int * outl,
+                       const unsigned char * in,
+                       int inl) {
+    OpensslCallIsOne::callChecked(lib::OpenSSLLib::SSL_EVP_CipherUpdate, ctx, outm, outl, in, inl);
+}
+
+void _EVP_CipherFinal_ex(EVP_CIPHER_CTX* ctx, unsigned char* outm, int* outl) {
+    OpensslCallIsOne::callChecked(lib::OpenSSLLib::SSL_EVP_CipherFinal_ex, ctx, outm, outl);
+}
+
+int _EVP_CIPHER_CTX_key_length(const EVP_CIPHER_CTX *ctx) {
+    return lib::OpenSSLLib::SSL_EVP_CIPHER_CTX_key_length(ctx);
+}
+
+int _EVP_CIPHER_CTX_iv_length(const EVP_CIPHER_CTX *ctx) {
+    int res = lib::OpenSSLLib::SSL_EVP_CIPHER_CTX_iv_length(ctx);
+    if (0 == res) {
+        throw OpenSSLException{"The cipher does not use an IV."};
```

Comment:
Do we really want to throw every time this is called on a cipher that doesn't support IVs? Wouldn't this be counter-productive for people that want to write code that works with multiple ciphers?

Reply:
Don't see a problem here. It is an internal API that is not used by end-users. Inside the library itself, we should check if the cipher supports a given operation before calling it.